sessions/2018/october/Session_15.md

```diff
@@ -0,0 +1,324 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 15 (16/10/2018)
+
+- Reviewing/fixing Spring API homework
+- API desing concepts
+- Dependency injection concepts
+
+### Notes:
+
+Roger was reviewing my homework solution and advicing me some modifications and improvements.
+
+#### Reviewing/fixing Spring API homework
+
+In certain part of the code I needed to save and incremental `id` `int` field into the object's field that was a `String`. A way to do it should be changing the type of the object's field, but a simpler way (hack) to work around it is doing `id` increment and mapping to `String` by concatenating with `""` at the moment to set into the object's field:
+
+```java
+book.setBookID(""+id++);
+```
+
+__Note__: In the final solution, the object field was changed from `String` to `Integer` (handle `null` values is convenient when answering to request to inexistents elements), since the previous mentioned hack was a workaround.
+
+Another optimization that was spoken was about to change this kind of return conditions when using `Map` structure:
+
+```java
+public Book getBook(String id) {
+    if (bookCache.get(id) != null){
+        return bookCache.get(id);
+    } else {
+        return new Book();
+    }
+}
+```
+
+To use the included `getOrDefault` method that comes integrated into `Map` structure:
+
+```java
+public Book getBook(String id) {
+    return bookCache.getOrDefault(id, new Book());
+    }
+```
+
+#### Use of the API application
+
+When we run the application (see note below), a Tomcat server is started on port `8080`.
+
+__Note__: If changes has been made, the application needs to be restarted. So it's important don't use `Ctrl+Shift+F10` (or the equivalent running by mouse click action), because this will start another Tomcat server instance. Instead of that, what we need to do is go over the bottom panel, and click over the green rerun arrow (first icon on the top of the left vertical menu), or just hit `Ctrl+5` key binding.
+
+__Note__: When we're designing [RESTful][1] interfaces, a main concept is we have to use nouns and not verbs, so `fibonacci_number` is a valid route name, while `calculate` isn't.
+
+##### GET Maps
+
+In order to use the API we have to access GET HTTP routes trought the browser, for example going to <http://localhost:8080/api/customer> we will receive the next HTTP response:
+
+```json
+[
+{
+name: "Test user 1",
```

Comment:
GitHub shows this in red because it's not actually JSON. Remember, you must have double quotes `""` around the key, i.e. the left-hand side of `:` in JSON.